objects.py

```diff
@@ -84,175 +81,368 @@ def __init__(self, fromjson = None):
         }
-        ret = switch[self.run_type](path)
-  #      os.chdir(orig)
+        ret = switch[self.run_type](path, handle)
+        #      os.chdir(orig)
         return ret
 
-    def _byteify(self, data):
-        # if this is a unicode string, return its string representation
-      #  #   return data.encode('utf-8')
-        # if this is a list of values, return list of byteified values
-        if isinstance(data, list) and len(data) == 1:
-            return self._byteify(data[0])
-        # if it's anything else, return it in its original form
-        return data
-
-    def run_composite(self, path):
-        opts = self._spec_opts()
+    def run_composite(self, path, handle):
+        """Called internally only by this.run()"""
+        cmd = '{} {} -jar {}/specjbb2015.jar -m COMPOSITE {}'.format(
+            self.jdk, self.jvm_options, path, self._spec_opts())
         for x in range(self.num_runs):
             result_dir = self._prerun(path)
-            #p = Popen([self.jdk, self.jvm_options, "-jar", "{}/specjbb2015.jar".format(path), "-m", "COMPOSITE", opts, "2>", "{}/composite.log".format(result_dir), "{}/composite.out".format(result_dir)])
-            #p.wait()
- #           spec = os.subprocess.Popen(['/usr/bin/java' "{}/specjbb2015.jar".format(path)])
-            os.system('{} {} -jar {}/specjbb2015.jar -m COMPOSITE {} 2> {}/composite.log > {}/composite.out &'.format(self.jdk, self.jvm_options, path, opts, result_dir, result_dir))
+            handle('Starting run number {}.'.format(x))
+            handle('Using command: "{}"'.format(cmd))
+            errout = open(os.path.join(result_dir, 'composite.out'), 'wb')
+            stdout = open(os.path.join(result_dir, 'composite.log'), 'wb')
+            p = Popen(shlex.split(cmd), cwd=path, stdout=PIPE, stderr=PIPE)
+            while (p.poll() is None):
+                e = p.stderr.readline()
+                if (e != ''):
+                    errout.write(e)
+                    handle(e)
+                o = p.stdout.readline()
+                if (o != ''):
+                    stdout.write(o)
+                    handle(o)
+            errout.close()
+            stdout.close()
+            exitcode = p.wait()
+            if (exitcode != 0):
+                return -1
+
         return 0
 
-    def run_distributed_ctrl_txl(self, path):
+    def run_distributed_ctrl_txl(self, path, handle):
+        """Called internally only by this.run()"""
         ctrl_ip = self.properties.root['specjbb.controller.host'].value
 
-        #Check if ip responds here
+        # Check if ip responds here
+        pingcmd = 'ping -c 1 {}'.format(ctrl_ip)
+        FNULL = open(os.devnull, 'w')
+        ping = Popen(shlex.split(pingcmd), stderr=FNULL, stdout=FNULL)
+        exitcode = ping.wait()
+        FNULL.close()
+        if (exitcode != 0):
+            handle('ERROR: Failed to ping Controller host (specjbb.controller.host): {}'.format(ctrl_ip))
+            return 4
```

Comment:
Would we want to return the popen'd exit code rather than our own? Is there any information in the code that we might want to pass to the user?